server: ensure drain of client connections at the end of decommission

cameronnunez opened the pull request:
Fixes #72754.

Previously, operators did not have the option to drain late into the decommissioning process, which caused queries to hang. This was a result of SQL client connections at the end of the process becoming suddenly unusable. This patch ensures that these client connections on the decommissioning node are closed before the node is marked as decommissioned. 

There is one caveat which is that the draining step of the decommissioning process will be skipped if the node serving the decommission request is a target. A follow-up PR will fix this by enforcing that nodes serving requests are not target nodes. Decommission requests will be routed to other nodes in the cluster that are not targets. 

### What's the difference between drain and decommission?

A drain prepares a node for process termination. This process moves all range leases away from the node but keeps the replicas. While a node is in the draining state, ranges cannot be rebalanced onto the node. 

Here are the steps in the drain sequence (**documentation needed**):

1. Initial wait
    - Continues normal node operation
    - Health check `/health?ready=1` starts returning false
    - LBs & connection management tools begin routing traffic to other nodes
    - Minimum duration: `server.shutdown.drain_wait` (default 0s)
            - This should be configured in coordination with the LB’s settings
2. ([DEV IN PROGRESS](https://github.com/cockroachdb/cockroach/pull/72991)) Wait for clients to close SQL connections
    - Maximum duration: `server.shutdown.connection_wait` (default 0s)
3. SQL connection closure
    - Completes processing of transactions started on the node
    - Refuses new connections and transactions
    - Finally, closes all SQL connections
    - Maximum duration: `server.shutdown.query_wait` (default 10s)
4. Lease and leadership transferral
    - Transfers all range leases and raft leaderships to other nodes
    - Maximum duration between iterations in the infinite loop: `server.shutdown.lease_transfer_wait` (default 5s)

A decommission prepares a node for permanent removal and **(NEW) process termination**. This process moves all range replicas away from the node.

Here are the steps in the decommission sequence (**documentation needed**):
1. Replica transferral
    - Transfers all range replicas to other nodes
2. **(NEW) Drain**
3. Final state transition
    - Marks the node as fully decommissioned


### How does shutdown work in relation to drain and decommission?

Shutdown consists of two phases: (1) shutdown preparation (with either a drain or decommission) and (2) process termination. 

If the goal is to have the node restart and continue to be active (e.g. software upgrade), the first phase should be a drain. On the other hand, if the node should be permanently removed from the cluster, the first phase should be a decommission; it is optional, but recommended, to precede the decommission with a drain to avoid possible disturbances in query performance. 

Once the first phase is complete, the process will still be running. It is only stopped after process termination. 


### Release notes


Release note (bug fix): `cockroach node decommission` no longer causes query failure due to the decommissioning node not closing open SQL connections and still being marked as ready. The decommissioning process now includes a draining step that fixes this. In other words, a decommission now automatically drains a node. This also means that running a drain after a decommission is no longer necessary. It is optional, but recommended, that `cockroach node drain` is used before `cockroach node decommission` to avoid the possibility of a disturbance in query performance. 

Release note (cli change): the flag `--self` of the `cockroach node decommission` command is deprecated. Instead, operators should specify the node ID of the target node as an explicit argument. The node that the command is connected to should not be a target node.

cameronnunez committed:
server: ensure drain of client connections at the end of decommission

Previously, operators did not have the option to drain late into the decommissioning process,
which caused queries to hang. This was a result of SQL client connections at the end of the
process becoming suddenly unusable. This patch ensures that these client connections on the
decommissioning node are closed before the node is marked as decommissioned.

There is one caveat which is that the draining step of the decommissioning process will be
skipped if the node serving the decommission request is a target. A follow-up PR will fix
this by enforcing that nodes serving requests are not target nodes. Decommission requests
will be routed to other nodes in the cluster that are not targets.

Release note (bug fix): `cockroach node decommission` no longer causes query failure due to
the decommissioning node not closing open SQL connections and still being marked as ready.
The decommissioning process now includes a draining step that fixes this. In other words,
a decommission now automatically drains a node. This also means that running a drain after
a decommission is no longer necessary. It is optional, but recommended, that
`cockroach node drain` is used before `cockroach node decommission` to avoid the
possibility of a disturbance in query performance.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/74319)
<!-- Reviewable:end -->

irfansharif reviewed (commented):
LGTM, but will leave the final rubber stamp to your team. This would benefit from a test, and a more verbose release note in the commit message itself. It'll help to compare against what the status quo was before this PR.

cameronnunez commented:
> LGTM, but will leave the final rubber stamp to your team. This would benefit from a test, and a more verbose release note in the commit message itself. It'll help to compare against what the status quo was before this PR.

Thanks for taking a look. Still a draft - currently working on the roachtest for this. And noted on the release notes, I'll be sure to update that.

irfansharif commented on pkg/cmd/roachtest/tests/decommission.go:
```diff
@@ -993,6 +1004,85 @@ func runDecommissionRandomized(ctx context.Context, t test.Test, c cluster.Clust
+		t.Fatal(err)
+	}
+
+	// Run queries while the node is decommissioning.
```
This comment needs updating.

irfansharif commented on pkg/cmd/roachtest/tests/decommission.go:
```diff
@@ -62,6 +62,17 @@ func registerDecommission(r registry.Registry) {
 		})
 	}
+	{
+		numNodes := 5
```
Is a 4 node cluster here sufficient?

irfansharif commented on pkg/cmd/roachtest/tests/decommission.go:
```diff
@@ -993,6 +1004,85 @@ func runDecommissionRandomized(ctx context.Context, t test.Test, c cluster.Clust
+
+	var (
+		pinnedNodeID = 5
+		pinnedNode   = c.Node(5)
```
[nit] Here and in the comments below, use `pinnedNodeID` instead of hardcoding the constant 5.

irfansharif commented on pkg/cli/node.go:
```diff
@@ -467,12 +466,28 @@ func runDecommissionNodeImpl(
+			time.Sleep(200 * time.Millisecond)
+
+			// Finally, mark the nodes as fully decommissioned.
+			decommissionReq := &serverpb.DecommissionRequest{
```
Are there issues if the CLI is issuing this RPC against the node that was just drained? Or does the drain process only expunge SQL connections?

knz reviewed (commented):
> Reviewed 3 of 3 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74319)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @cameronnunez and @irfansharif)

___
*[-- commits, line 4 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74319#-MtEZLSX7L4k2SUbZNVU):*
This needs a sufficiently elaborate release note.
___
*[pkg/cli/node.go, line 476 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74319#-MtEZYFq0kz5NcdUpMJC:-MtEZYFq0kz5NcdUpMJD:bqzzat) ([raw file](https://github.com/cockroachdb/cockroach/blob/93a7431cc8c2cccd9e6bbd8d4466d3e0a02cf406/pkg/cli/node.go#L476)):*
> ```Go
> 					NodeId:   targetNode.String(),
> 				}
> 				if _, err = c.Drain(ctx, drainReq); err != nil {
> ```

This needs a comment to explain why it's not necessary to iterate on `Drain()`, like is done in the other places.
___
*[pkg/cli/node.go, line 478 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74319#-MtEZdhO9aLBuEXBwhXG:-MtEZdhO9aLBuEXBwhXH:b88ci9v) ([raw file](https://github.com/cockroachdb/cockroach/blob/93a7431cc8c2cccd9e6bbd8d4466d3e0a02cf406/pkg/cli/node.go#L478)):*
> ```Go
> 				if _, err = c.Drain(ctx, drainReq); err != nil {
> 					fmt.Fprintln(stderr)
> 					return errors.Wrapf(err, "while trying to drain %s", targetNode.String())
> ```

`... while trying to drain n%d", targetNode)`
___
*[pkg/cli/node.go, line 482 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74319#-MtEZjOvBtg9zXs85Xmk:-MtEZjOvBtg9zXs85Xml:buqru8v) ([raw file](https://github.com/cockroachdb/cockroach/blob/93a7431cc8c2cccd9e6bbd8d4466d3e0a02cf406/pkg/cli/node.go#L482)):*
> ```Go
> 			}
> 
> 			// Allow time for the drains to close the client connections.
> ```

Can you explain why this is needed?
___
*[pkg/cli/node.go, line 486 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74319#-Ms8MpE1_lmNLbG3uMiH-r1-486:-MtEZn7YB3UXZ8Dropcl:b-wjz6ep) ([raw file](https://github.com/cockroachdb/cockroach/blob/93a7431cc8c2cccd9e6bbd8d4466d3e0a02cf406/pkg/cli/node.go#L486)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

Are there issues if the CLI is issuing this RPC against the node that was just drained? Or does the drain process only expunge SQL connections?
</blockquote></details>

Yeah good point. I think _all_ the decommission/drain requests need to be routed via a node that's not being targeted by the command.

It may be useful to assert at the very beginning (at the top of the deco function) that the node ID of the node serving the adminClient (`c`  here) is not included in `targetNodes`.
___
*[pkg/cli/quit.go, line 177 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74319#-MtE_4q2-OmPtJ503Tr1:-MtE_4q3A6KZ82Dx_M1T:b-n1snd6) ([raw file](https://github.com/cockroachdb/cockroach/blob/93a7431cc8c2cccd9e6bbd8d4466d3e0a02cf406/pkg/cli/quit.go#L177)):*
> ```Go
> 				// Unexpected error.
> 				fmt.Fprintf(stderr, "\n") // finish the line started above.
> 				log.Infof(ctx, "graceful drain failed: %v", err)
> ```

It's good to change this, but that deserves a separate commit.
___
*[pkg/cmd/roachtest/tests/decommission.go, line 1012 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74319#-MtE_nXq2FQHFzgpfXYD:-MtE_nXq2FQHFzgpfXYE:brhbuo2) ([raw file](https://github.com/cockroachdb/cockroach/blob/93a7431cc8c2cccd9e6bbd8d4466d3e0a02cf406/pkg/cmd/roachtest/tests/decommission.go#L1012)):*
> ```Go
> // requests should continue to run without fail, even after the final range has been moved away
> // from the decommissioned node.
> func runDecommissionDrains(ctx context.Context, t test.Test, c cluster.Cluster, numNodes int) {
> ```

TBH I'm not exactly sure what this test is doing wrt the SQL clients.
The previous behavior is that a SQL client was able to connect successfully, but then became unable to perform SQL queries.

We are OK if a SQL client gets "Connection refused" after a node has finished draining. That connection error should not incur a test failure.

So we want in the test a condition that looks like this:
1. do a SQL connect . If connection fails -> Stop, but no test error
2. using the open SQL connection, send a query. If the query fails -> Stop with a test error
___
*[pkg/cmd/roachtest/tests/decommission.go, line 1013 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74319#-MtE_ENe0HE5uujpuzG2:-MtE_ENe0HE5uujpuzG3:b-24ferq) ([raw file](https://github.com/cockroachdb/cockroach/blob/93a7431cc8c2cccd9e6bbd8d4466d3e0a02cf406/pkg/cmd/roachtest/tests/decommission.go#L1013)):*
> ```Go
> // from the decommissioned node.
> func runDecommissionDrains(ctx context.Context, t test.Test, c cluster.Cluster, numNodes int) {
> ```

nit: no empty space at start of function body
___
*[pkg/cmd/roachtest/tests/decommission.go, line 1015 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74319#-MtE_Yhs6zodtpabafh-:-MtE_Yhs6zodtpabafh0:b-l5w8ni) ([raw file](https://github.com/cockroachdb/cockroach/blob/93a7431cc8c2cccd9e6bbd8d4466d3e0a02cf406/pkg/cmd/roachtest/tests/decommission.go#L1015)):*
> ```Go
> 
> 	var (
> 		pinnedNodeID = 5
> ```

What does "pinned" mean?
___
*[pkg/cmd/roachtest/tests/decommission.go, line 1064 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74319#-MtE_QHp2e16RbJW2tt2:-MtE_QHp2e16RbJW2tt3:b-2chg51) ([raw file](https://github.com/cockroachdb/cockroach/blob/93a7431cc8c2cccd9e6bbd8d4466d3e0a02cf406/pkg/cmd/roachtest/tests/decommission.go#L1064)):*
> ```Go
> 
> 	if err := retry.WithMaxAttempts(ctx, retryOpts, maxAttempts, func() error {
> 		o, err := h.decommission(ctx, pinnedNode, pinnedNodeID, "--wait=none", "--format=csv")
> ```

See my comments earlier: I think that the _target_ node, and the node that's serving the deco request should be different.
___
*[pkg/cmd/roachtest/tests/decommission.go, line 1070 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74319#-MtE_UY34r1uuqsZHPNJ:-MtE_UY34r1uuqsZHPNK:b-f9aaxx) ([raw file](https://github.com/cockroachdb/cockroach/blob/93a7431cc8c2cccd9e6bbd8d4466d3e0a02cf406/pkg/cmd/roachtest/tests/decommission.go#L1070)):*
> ```Go
> 
> 		// Run queries while the node is decommissioning.
> 		e := c.RunE(ctx, pinnedNode, `./cockroach sql -e "SHOW DATABASES" --insecure`)
> ```

The SQL client should connect to the node that's the target of the decommission.


<!-- Sent from Reviewable.io -->

knz commented:
A reminder: a test for a bug fix should be checked to fail on a branch without the fix, and succeed with the fix.

If you only check the test passes with the fix, that doesn't say much. you could have an empty test and it would also pass with the fix.

cameronnunez commented:
> A reminder: a test for a bug fix should be checked to fail on a branch without the fix, and succeed with the fix.
> 
> If you only check the test passes with the fix, that doesn't say much. you could have an empty test and it would also pass with the fix.

The test is still incomplete and does not satisfy what you describe here, which is why I still have it in the draft state. Should have written WIP in the title however.

cameronnunez commented:
To be clear, this PR wasn't ready for the review cycle quite yet, but appreciate the early feedback

cameronnunez reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74319)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @cameronnunez, @irfansharif, and @knz)

___
*[pkg/cli/node.go, line 476 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74319#-MtEZYFq0kz5NcdUpMJC:-MtFIE7D9pOtV1qs0mzn:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/93a7431cc8c2cccd9e6bbd8d4466d3e0a02cf406/pkg/cli/node.go#L476)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

This needs a comment to explain why it's not necessary to iterate on `Drain()`, like is done in the other places.
</blockquote></details>

Done.
___
*[pkg/cli/node.go, line 478 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74319#-MtEZdhO9aLBuEXBwhXG:-MtFIEwYD565jbS-ISl9:bt2vr1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/93a7431cc8c2cccd9e6bbd8d4466d3e0a02cf406/pkg/cli/node.go#L478)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

`... while trying to drain n%d", targetNode)`
</blockquote></details>

Fixed.
___
*[pkg/cli/node.go, line 482 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74319#-MtEZjOvBtg9zXs85Xmk:-MtFG3uM7fAP_ki8P_0J:b-dpdh3x) ([raw file](https://github.com/cockroachdb/cockroach/blob/93a7431cc8c2cccd9e6bbd8d4466d3e0a02cf406/pkg/cli/node.go#L482)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Can you explain why this is needed?
</blockquote></details>

Don't necessarily think it is. I was just experimenting when constructing the test (currently in draft state still). 
___
*[pkg/cli/node.go, line 486 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74319#-Ms8MpE1_lmNLbG3uMiH-r1-486:-MtFGPWK1l1um2e5LAiv:b7p6q0) ([raw file](https://github.com/cockroachdb/cockroach/blob/93a7431cc8c2cccd9e6bbd8d4466d3e0a02cf406/pkg/cli/node.go#L486)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Yeah good point. I think _all_ the decommission/drain requests need to be routed via a node that's not being targeted by the command.

It may be useful to assert at the very beginning (at the top of the deco function) that the node ID of the node serving the adminClient (`c`  here) is not included in `targetNodes`.
</blockquote></details>

So this would mean the "--self" flag is a blocker, correct? removal of this would be a breaking change?
___
*[pkg/cli/quit.go, line 177 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74319#-MtE_4q2-OmPtJ503Tr1:-MtFGhCg8eDeXz7AmYRK:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/93a7431cc8c2cccd9e6bbd8d4466d3e0a02cf406/pkg/cli/quit.go#L177)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

It's good to change this, but that deserves a separate commit.
</blockquote></details>

Done.
___
*[pkg/cmd/roachtest/tests/decommission.go, line 66 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74319#-Mt3rKacwE1WaJ5qrUlM-r1-66:-MtFGiaz0yuYqp9v4wyZ:b-b9ytw) ([raw file](https://github.com/cockroachdb/cockroach/blob/93a7431cc8c2cccd9e6bbd8d4466d3e0a02cf406/pkg/cmd/roachtest/tests/decommission.go#L66)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

Is a 4 node cluster here sufficient?
</blockquote></details>

Yes it is. Will do this instead. 
___
*[pkg/cmd/roachtest/tests/decommission.go, line 1012 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74319#-MtE_nXq2FQHFzgpfXYD:-MtFGmsV2Iwzljb8CUnj:byzwlko) ([raw file](https://github.com/cockroachdb/cockroach/blob/93a7431cc8c2cccd9e6bbd8d4466d3e0a02cf406/pkg/cmd/roachtest/tests/decommission.go#L1012)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

TBH I'm not exactly sure what this test is doing wrt the SQL clients.
The previous behavior is that a SQL client was able to connect successfully, but then became unable to perform SQL queries.

We are OK if a SQL client gets "Connection refused" after a node has finished draining. That connection error should not incur a test failure.

So we want in the test a condition that looks like this:
1. do a SQL connect . If connection fails -> Stop, but no test error
2. using the open SQL connection, send a query. If the query fails -> Stop with a test error


</blockquote></details>

My comment was poorly worded. Should read "queries should not fail." Fixed. 
___
*[pkg/cmd/roachtest/tests/decommission.go, line 1015 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74319#-MtE_Yhs6zodtpabafh-:-MtFH48DCZgm7JppqG22:bthottb) ([raw file](https://github.com/cockroachdb/cockroach/blob/93a7431cc8c2cccd9e6bbd8d4466d3e0a02cf406/pkg/cmd/roachtest/tests/decommission.go#L1015)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

What does "pinned" mean?
</blockquote></details>

Poor naming on my part. Changing to "decomm" to represent the node that is to be decommissioned. Fixed. Now, "pinned" will refer to the node that's serving the decommission request.
___
*[pkg/cmd/roachtest/tests/decommission.go, line 1016 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74319#-Mt3rKacwE1WaJ5qrUlM-r1-1016:-MtFHdow3AAxwVjGosKh:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/93a7431cc8c2cccd9e6bbd8d4466d3e0a02cf406/pkg/cmd/roachtest/tests/decommission.go#L1016)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

[nit] Here and in the comments below, use `pinnedNodeID` instead of hardcoding the constant 5. 
</blockquote></details>

Done.
___
*[pkg/cmd/roachtest/tests/decommission.go, line 1064 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74319#-MtE_QHp2e16RbJW2tt2:-MtFHytE5Rgu9BJm7rwS:bt2vr1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/93a7431cc8c2cccd9e6bbd8d4466d3e0a02cf406/pkg/cmd/roachtest/tests/decommission.go#L1064)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

See my comments earlier: I think that the _target_ node, and the node that's serving the deco request should be different.
</blockquote></details>

Fixed.
___
*[pkg/cmd/roachtest/tests/decommission.go, line 1070 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74319#-MtE_UY34r1uuqsZHPNJ:-MtFI12K9Ya94NQl782R:barbgbw) ([raw file](https://github.com/cockroachdb/cockroach/blob/93a7431cc8c2cccd9e6bbd8d4466d3e0a02cf406/pkg/cmd/roachtest/tests/decommission.go#L1070)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

The SQL client should connect to the node that's the target of the decommission.
</blockquote></details>

Changed the variable name to be more clear. See other comment above.
___
*[pkg/cmd/roachtest/tests/decommission.go, line 1080 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74319#-Mt3rKacwE1WaJ5qrUlM-r1-1080:-MtFIC9lAeeN2QqOM-eA:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/93a7431cc8c2cccd9e6bbd8d4466d3e0a02cf406/pkg/cmd/roachtest/tests/decommission.go#L1080)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

This comment needs updating.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

cameronnunez commented:
It seems the removal of `--self` is a blocker, since we need the node serving the the decommission/drain requests to not be a target node. And this would be a breaking change, right?

knz commented:
As we discussed we can also imagine a mode where the tool searches for another node to serve as "control" node.

cameronnunez commented:
Right. In that scenario, issuing `decommission --self` against a node that was just drained should still succeed. Another node will be chosen as "control node" for decommissioning the drained node.

irfansharif commented:
We could deprecate --self, but we can't remove it in 22.1 for backwards compatibility reasons. In the short term instead of finding another node to route the requests through (assuming it's more code/harder to get right), we could just skip the drain step if --self was used. Again, logging loudly that it's deprecated.

cameronnunez committed:
server: warn that `--self` is deprecated

Release note (cli change): the flag `--self` of the `cockroach node decommission`
command is deprecated. Instead, operators should specify the node ID of the target
node as an explicit argument. The node that the command is connected to should not be
a target node.

cameronnunez committed:
server: fix comment in `quit.go`

Release note: None

cameronnunez commented:
For this PR, as suggested by Irfan, the drain will be skipped if the node serving the request is a target node. So, all SQL connections will be drained by the end of decommission, except in the edge case where the node is the host and also a target node. Another PR will be opened to handle this case, by routing the decommission request to another node in the cluster.

cameronnunez commented:
RFAL

irfansharif commented on pkg/cli/node.go:
```diff
@@ -467,12 +470,34 @@ func runDecommissionNodeImpl(
+			for _, targetNode := range nodeIDs {
+				if targetNode == serverCfg.IDContainer.Get() {
+					// Skip the draining step for the node serving the request, if it is a target node.
+					fmt.Fprintf(stderr, ""+
```
[nit] Do we need the `""+`. Instead of "this node", perhaps include the node ID itself given we have it?

cameronnunez replied:
Fixed.

irfansharif commented on pkg/cli/node.go:
```diff
@@ -431,12 +435,11 @@ func runDecommissionNodeImpl(
+	// 1) Mark each node as 'decommissioning'. In doing so, replicas are slowly moved off of these nodes.
+	// 2) After when we're made aware that the replica counts have all hit zero, and that all nodes
+	// have been successfully marked as 'decommissioning', drain each node.
+	// 3) After each node has been drained, mark each node as 'decommissioned'.
```
Maybe a few words here about if the node we're talking to is the one being decommissioned, and why that's frowned upon+skips the drain step? Also [nit] wrap these comments at 80 chars.

cameronnunez replied:
Done.

irfansharif commented on pkg/cli/node.go:
```diff
@@ -316,8 +316,12 @@ func runDecommissionNode(cmd *cobra.Command, args []string) error {
 	defer cancel()
 
+	if nodeCtx.nodeDecommissionSelf {
+		fmt.Fprintf(stderr, "warning: --%s is deprecated.\n", cliflags.NodeDecommissionSelf.Name)
```
[nit] Maybe just a few more words here too about how we're going to skip the drain step.

cameronnunez replied:
That warning happens here so that this warning is also sent in the more general case where a target node is the host:

```go 
// We now drain the nodes in order to close all SQL connections.
// Note: iteration is not necessary here since there are no remaining leases
// on the decommissioning node after replica transferral.
for _, targetNode := range nodeIDs {
  if targetNode == localNodeID {
    // Skip the draining step for the node serving the request, if it is a target node.
    log.Warning(ctx,
      "node specified with --host should not be a target node; skipped the drain step for this node.",
    )
    continue
  }
  .
  .
  .
}

irfansharif commented on pkg/cmd/roachtest/tests/decommission.go:
```diff
@@ -993,6 +1006,94 @@ func runDecommissionRandomized(ctx context.Context, t test.Test, c cluster.Clust
+		o, e := h.decommission(ctx, decommNode, pinnedNodeID, "--wait=none", "--format=csv")
+		require.NoError(t, errors.Wrapf(e, "decommission failed: %v"))
+
+		// Near the end of the decommission, check if the SQL connection is still open.
```
To simplify this code block a bit, how about:
```go
if err := cli.MatchCSV(o, expDecommissioning); err != nil {
  return err
}

// Still not done decommissioning; check to see if we're drained.
if drained := db.Ping() == nil; drained {
    return nil
}

// We haven't been decommissioned or drained, make sure SQL connections are functional.
queryErr := run(db, `SHOW DATABASES`)
require.NoError(t, queryErr)
return nil
```

I'm not fully following what the 5s retry achieves. Mind explaining?

cameronnunez replied:
👍 Used your code block as inspiration to simplify it a bit. It doesn't quite work as it but made changes so it does.

irfansharif reviewed (commented):
Pretty slick! Just a few comments, though as before, all this would definitely benefit from Rafa's input.

cameronnunez commented:
RFAL

irfansharif commented on pkg/cli/node.go:
```diff
@@ -467,12 +477,34 @@ func runDecommissionNodeImpl(
+				if targetNode == localNodeID {
+					// Skip the draining step for the node serving the request, if it is a target node.
+					log.Warning(ctx,
+						"node specified with --host should not be a target node; skipped the drain step for this node.",
```
It's not necessary that we hit this code path through --host, right? We could use --self and rely on the default --host value? I'd make this warning a bit more generic perhaps (and also avoid the scary "should"): 

"skipping drain step for node %d, decommissioning through the node itself" or something similar.

cameronnunez replied:
Done.

irfansharif commented on pkg/cli/node.go:
```diff
@@ -467,12 +477,34 @@ func runDecommissionNodeImpl(
+					NodeId:   targetNode.String(),
+				}
+				if _, err = c.Drain(ctx, drainReq); err != nil {
+					fmt.Fprintln(stderr)
```
Was this empty println intentional?

cameronnunez replied:
Yes, this pattern is present throughout the function originally, I think to ensure that the returned error starts on a new line once it's eventually written to `stderr`.

irfansharif commented on pkg/cli/node.go:
```diff
@@ -559,7 +591,13 @@ func runRecommissionNode(cmd *cobra.Command, args []string) error {
 	s := serverpb.NewStatusClient(conn)
 
-	nodeIDs, err = handleNodeDecommissionSelf(ctx, s, nodeIDs, "recommissioning")
+	r, err := s.Node(ctx, &serverpb.NodeRequest{NodeId: "local"})
```
[nit] This block used here and above could be a tiny helper, "getLocalNodeID" perhaps?

cameronnunez replied:
Done.

irfansharif reviewed (approved):
LGTM, this is pretty great! And thanks for writing the roachtest as well. I'd run it manually a few times before merging, just to shake out any possible flakes we missed (looks robust to me though). Left a few nits below.

knz reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> with nit

> Reviewed 2 of 5 files at r10, 1 of 3 files at r13, 1 of 1 files at r14, 1 of 1 files at r15, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74319#-:-MulE6Ng8WBOYB0wHtH9:b-xv64lu)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @aayushshah15, @cameronnunez, and @irfansharif)

___
*[-- commits, line 14 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/74319#-MulBPjVCrNB6sAJxeud):*
"no longer causes query failure."
with a period at the end this suggests that there won't be query failures ever again, in any circumstances. I think that's a bit too bold of a statement.

It's customary for bug fixes to clarify what was the condition where the problem was occurring. Something like: "no longer causes query failure when ...."
(you can fill in the rest)
___
*[pkg/cli/node.go, line 497 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/74319#-Ms8MpE1_lmNLbG3uMiH-r12-497:-MulCjno4rtreTqE52nP:b88j36h) ([raw file](https://github.com/cockroachdb/cockroach/blob/c289e5ed391c070d4e6b8f6f2dc0ff4e80670b8c/pkg/cli/node.go#L497)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

Was this empty println intentional?
</blockquote></details>

I have the same question.
___
*[pkg/cli/node.go, line 497 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/74319#-MulCUci7FI1rdxMbIpx:-MulCUci7FI1rdxMbIpy:bbttjwz) ([raw file](https://github.com/cockroachdb/cockroach/blob/ac8a0853c3afbc092439afafe19d1f9301f14f2c/pkg/cli/node.go#L497)):*
> ```Go
> 					log.Warningf(ctx,
> 						"skipping drain step for node %s; it is decommissioning and serving the request",
> 						localNodeID.String(),
> ```

please format with `node n%d`
and then pass `localNodeID` as argument without calling String() on it.


<!-- Sent from Reviewable.io -->

aayushshah15 reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> I don't have anything aside from the suggestions @irfansharif already expressed.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74319#-:-MukwMkzEZbihi3nNehB:br7e1m0)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @cameronnunez and @irfansharif)



<!-- Sent from Reviewable.io -->

knz commented:
oh I forgot please add in the commit message:

```
Release note (cli change): the `--self` flag to `cockroach node decommission` is now deprecated. Instead,
operators are invited to specify the node ID of the target node as explicit argument. The target node ID
should preferably be another node than the one that the command is connected to.
```

cheers

jtsiros commented:
cc: @taroface

cameronnunez commented on pkg/cmd/roachtest/tests/decommission.go:
```diff
@@ -993,6 +1006,100 @@ func runDecommissionRandomized(ctx context.Context, t test.Test, c cluster.Clust
+		require.NoError(t, errors.Wrapf(e, "decommission failed"))
+
+		// Check if all the replicas have been transferred.
+		if e := cli.MatchCSV(o, expReplicasTransferred); e != nil {
```
Done.

cameronnunez commented on pkg/cmd/roachtest/tests/decommission.go:
```diff
@@ -993,6 +1006,100 @@ func runDecommissionRandomized(ctx context.Context, t test.Test, c cluster.Clust
+
+		// Check to see if the node has been drained.
+		// If not, send a query and verify that the query does not fail.
+		if drained := db.Ping() != nil; !drained {
```
Done.

cameronnunez commented on pkg/cmd/roachtest/tests/decommission.go:
```diff
@@ -993,6 +1006,100 @@ func runDecommissionRandomized(ctx context.Context, t test.Test, c cluster.Clust
+	// Check that the node is fully decommissioned.
+	o, err := h.decommission(ctx, decommNode, pinnedNodeID, "--format=csv")
+	require.NoError(t, err)
+	err = cli.MatchCSV(o, expDecommissioned)
```
Done.

cameronnunez commented on pkg/cmd/roachtest/tests/decommission.go:
```diff
@@ -993,6 +1006,100 @@ func runDecommissionRandomized(ctx context.Context, t test.Test, c cluster.Clust
+	}
+
+	// Connect to node 4 (the target node of the decommission).
+	db := c.Conn(ctx, t.L(), decommNodeID)
```
Done.

cameronnunez commented:
> oh I forgot please add in the commit message:
> 
> ```
> Release note (cli change): the `--self` flag to `cockroach node decommission` is now deprecated. Instead,
> operators are invited to specify the node ID of the target node as explicit argument. The target node ID
> should preferably be another node than the one that the command is connected to.
> ```
> 
> cheers

To be clear do you mean update the release note? there exists a release note detailing this

cameronnunez commented:
TFTRs! 

bors r=knz,irfansharif,aayushshah15
